include/ignition/gazebo/components/Factory.hh

```diff
@@ -245,19 +267,32 @@ namespace components
       return comp;
     }
 
+    /// \brief Create a new instance of a component, initialized with particular
+    /// data.
+    /// \param[in] _type Component id to create.
+    /// \param[in] _data The data to populate the component instance with.
+    /// \return Pointer to a component. Null if the component
+    /// type could not be handled.
+    public: std::unique_ptr<components::BaseComponent> New(
+        const ComponentTypeId &_type, const components::BaseComponent *_data)
```

Comment:
Can we add some unit tests for this function in `ComponentFactory_TEST`? I'm mainly concerned about cases when the data passed in doesn't match `ComponentTypeId`.

---

As a side note, I'm trying to understand whether the `New` / `Create` functions that receive a raw `BaseComponent` pointer are being added out of necessity or out of convenience. There are all kinds of issues that may happen when mixing types, so we should be extra careful with those new APIs.

Reply:
Unit tests have been added in https://github.com/ignitionrobotics/ign-gazebo/pull/856/commits/d871df3ea4b941bfdec22c2aa8c1df7955a529fa

> As a side note, I'm trying to understand whether the `New` / `Create` functions that receive a raw `BaseComponent` pointer are being added out of necessity or out of convenience

They are added out of necessity (I think :sweat_smile:) I explained why the `Create` function was needed in https://github.com/ignitionrobotics/ign-gazebo/pull/856#discussion_r684550937. The `New` function is used here, for example:

https://github.com/ignitionrobotics/ign-gazebo/blob/660268e70d79899563cf63bdbc114b005a311e32/src/EntityComponentManager.cc#L644-L645

As you can see, we are creating a component and setting its data in a non-templated method. This is possible because `New` has access to the `BaseComponent`'s derived class's `data` via `Create`, which is templated (see https://github.com/ignitionrobotics/ign-gazebo/pull/856#discussion_r684550937).

Reply:
Ok, I think I understand now. It's all because we don't have the component storage classes anymore to create them for us, I believe. Thanks!